invokeai/backend/flux/trajectory_guidance_extension.py

```diff
@@ -0,0 +1,125 @@
+                A value of 0.0 is equivalent to vanilla image-to-image. A value of 1.0 will guide the denoising process
+                very close to the original latents.
+        """
+        assert 0.0 <= trajectory_guidance_strength <= 1.0
+        self._init_latents = init_latents
+        self._trajectory_guidance_strength = trajectory_guidance_strength
+        if inpaint_mask is None:
+            # The inpaing mask is None, so we initialize a mask with a single value of 1.0.
+            # This value will be broadcasted and treated as a mask of all 1s.
+            self._inpaint_mask = torch.ones(1, device=init_latents.device, dtype=init_latents.dtype)
+        else:
+            self._inpaint_mask = inpaint_mask
+
+    def _apply_mask_gradient_adjustment(self, t_prev: float) -> torch.Tensor:
+        """Applies inpaint mask gradient adjustment and returns the inpaint mask to be used at the current timestep."""
+        # As we progress through the denoising process, we promote gradient regions of the mask to have a full weight of
+        # 1.0. This helps to produce more coherent seams around the inpainted region. We experimented with a (small)
+        # number of promotion strategies (e.g. gradual promotion based on timestep), but found that a simple cutoff
+        # threshold worked well.
+        # We use a small epsilon to avoid any potential issues with floating point precision.
+        eps = 1e-4
+        mask_gradient_t_cutoff = 0.5
+        if t_prev > mask_gradient_t_cutoff:
+            # Early in the denoising process, use the inpaint mask as-is.
+            return self._inpaint_mask
+        else:
+            # After the cut-off, promote all non-zero mask values to 1.0.
+            mask = self._inpaint_mask.where(self._inpaint_mask <= (0.0 + eps), 1.0)
+
+        return mask
+
+    def step(
+        self, t_curr_latents: torch.Tensor, pred_noise: torch.Tensor, t_curr: float, t_prev: float
+    ) -> torch.Tensor:
+        # Handle gradient cutoff.
+        mask = self._apply_mask_gradient_adjustment(t_prev)
+
+        # Calculate the change_ratio based on the trajectory_guidance_strength.
+        # These mappings from trajectory_guidance_strength have no theoretical basis - they were tuned manually.
+        change_ratio_at_t_1 = build_line(x1=0.0, y1=1.0, x2=1.0, y2=0.0)(self._trajectory_guidance_strength)
+        change_ratio_at_cutoff = 1.0
+        t_cutoff = build_line(x1=0.0, y1=1.0, x2=1.0, y2=0.5)(self._trajectory_guidance_strength)
+        change_ratio = 1.0
+        if t_prev > t_cutoff:
+            # If we are before the cutoff, linearly interpolate between the change_ratio at t=1.0 and the change_ratio
+            # at the cutoff.
+            change_ratio = build_line(x1=1.0, y1=change_ratio_at_t_1, x2=t_cutoff, y2=change_ratio_at_cutoff)(t_prev)
+
+        mask = mask * change_ratio
```

Comment:
Does this need to be clamped

Reply:
It shouldn't need to be clamped, but I've added an assertion to verify that it is in the expected range so we can't screw it up: [cd3a7bd](https://github.com/invoke-ai/InvokeAI/pull/6900/commits/cd3a7bdb5ee3a37265e7c426bd9fbc88d95c5863)